### PR TITLE
Feature: support N configs -> 1 package scheme for Visual Studio generator

### LINKED ISSUE #2721: "visual_studio" generator support for multi-config variables (N configs -> 1 package)
I'm introducing ```conan``` in products of my company.
But, I found one limitation and I want improvement of it.

http://docs.conan.io/en/latest/creating_packages/package_approaches.html#n-configs-1-package

> The automatic conversion of multi-config variables to generators is currently only implemented in the cmake and txt generators. If you want to have support for them in another build system, please open a GitHub issue for it.

I found that sentence in above documentation. And I want to have support for them in **Visual Studio**, So I open this issue.

Thank you.

### PR DESCRIPTION
**Pull request by SSE4**
closes #2721 

Changelog: Feature: support N configs -> 1 package scheme for Visual Studio generator
Docs: https://github.com/conan-io/docs/pull/XXXX
tags: @slow, @svn

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

/cc @taeguk

### CODE REVIEW

**Comment by jgsogo** [comment]
I find this PR quite clear and interesting, but I will add a live test too to check that Visual Studio is parsing the file properly and taking into account the conditions. I would add a test running Visual Studio, are there are with CMake (search `"multi_config_files"`) or with QMake.

**Comment by lasote** [comment]
I agree. Please @SSE4 add a test using vs.

**Comment by SSE4** [comment]
@jgsogo @lasote I've added unit test that uses real VS

**Comment by SSE4** [comment]
@lasote I am not sure why does it fail, seems like unrelated tests:
```
======================================================================

FAIL: test_basic (blocked_v2.conans.test.functional.command.export_linter_test.ExportLinterTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/tmp/source/9b22/py36/blocked_v2/conans/test/functional/command/export_linter_test.py", line 37, in test_basic

    self._check_linter(client.user_io.out)

  File "/tmp/source/9b22/py36/blocked_v2/conans/test/functional/command/export_linter_test.py", line 44, in _check_linter

    self.assertIn("WARN: Linter. Line 8: Unused variable 'k'", output)

AssertionError: "WARN: Linter. Line 8: Unused variable 'k'" not found in [HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'url'. It is recommended to add it as attribute

[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'license'. It is recommended to add it as attribute

[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'description'. It is recommended to add it as attribute

Exporting package recipe

Python 3 incompatibilities

    ERROR: Py3 incompatibility. Line 8: Calling a dict.iter*() method

Linter warnings

    WARN: Linter. Line 8: Instance of &#x27;dict&#x27; has no &#x27;iteritems&#x27; member

    WARN: Linter. Line 8: Unused variable &#x27;k&#x27;

    WARN: Linter. Line 8: Unused variable &#x27;v&#x27;

Hello/1.2@lasote/stable: A new conanfile.py version was exported

Hello/1.2@lasote/stable: Folder: /tmp/tmpu56lxfwcconans/path with spaces/.conan/data/Hello/1.2/lasote/stable/export





======================================================================

FAIL: test_warning_as_errors (blocked_v2.conans.test.functional.command.export_linter_test.ExportLinterTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/tmp/source/9b22/py36/blocked_v2/conans/test/functional/command/export_linter_test.py", line 138, in test_warning_as_errors

    self._check_linter(client.user_io.out)

  File "/tmp/source/9b22/py36/blocked_v2/conans/test/functional/command/export_linter_test.py", line 44, in _check_linter

    self.assertIn("WARN: Linter. Line 8: Unused variable 'k'", output)

AssertionError: "WARN: Linter. Line 8: Unused variable 'k'" not found in [HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'url'. It is recommended to add it as attribute

[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'license'. It is recommended to add it as attribute

[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'description'. It is recommended to add it as attribute

Exporting package recipe

Python 3 incompatibilities

    ERROR: Py3 incompatibility. Line 8: Calling a dict.iter*() method

Linter warnings

    WARN: Linter. Line 8: Instance of &#x27;dict&#x27; has no &#x27;iteritems&#x27; member

    WARN: Linter. Line 8: Unused variable &#x27;k&#x27;

    WARN: Linter. Line 8: Unused variable &#x27;v&#x27;

ERROR: Package recipe has linter errors. Please fix them.
```

**Comment by jgsogo** [comment]
Have a look a this PR https://github.com/conan-io/conan/pull/4627, it fixes those tests, version `pylint==2.3.0` had a bug. Just merge changes from `develop` and it should be fixed.

**Comment by lasote** [comment]
We would like to try this with some patience, the release is already very delayed. I'm moving this to the next one. Probably we make a short and quick release so for the end of the month hopefully we can merge this,

File: conans/test/functional/generators/visual_studio_multi_test.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+void hello_debug()
+{{
+    std::cout << "Hello Debug" << std::endl;
+    print_arch();
+}}
+"""
+
+
+class VisualStudioMultiTest(unittest.TestCase):
+    @parameterized.expand([("visual_studio", "conanbuildinfo.props"),
```
Comment (memsharded) [inline conans/test/functional/generators/visual_studio_multi_test.py]: I don't understand this test very well.
Different output of Release/Debug and architecture is present in the source files.
But the output is never checked.

File: conans/test/functional/generators/visual_studio_multi_test.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                           ("visual_studio_multi", "conanbuildinfo_multi.props")])
+    @attr('slow')
+    @unittest.skipUnless(platform.system() == "Windows", "Requires MSBuild")
+    def build_vs_project_test(self, generator, props):
+        client = TestClient()
+        files = get_vs_project_files()
+        files["MyProject/main.cpp"] = main_cpp
+        files["conanfile.txt"] = conanfile_txt.format(generator=generator)
+        files["hello/conanfile.py"] = hello_conanfile_py
```
Comment (memsharded) [inline conans/test/functional/generators/visual_studio_multi_test.py]: Please do first the export (or create) of hello, exclusively.

Then later, define your consumer MyProject project.

File: conans/test/functional/generators/visual_studio_multi_test.py (hunk 1 above)
Comment (SSE4) [reply conans/test/functional/generators/visual_studio_multi_test.py]: you asked to add unit test that uses real Visual Studio to check that props files generated by conan could be parsed by Visual Studio, right?

File: conans/test/functional/generators/visual_studio_multi_test.py (hunk 2 above)
Comment (SSE4) [reply conans/test/functional/generators/visual_studio_multi_test.py]: I am not sure I understand this, export is done below

File: conans/test/functional/generators/visual_studio_multi_test.py (hunk 1 above)
Comment (SSE4) [reply conans/test/functional/generators/visual_studio_multi_test.py]: from @jgsogo :
```
I find this PR quite clear and interesting, but I will add a live test too to check that Visual Studio is parsing the file properly and taking into account the conditions. I would add a test running Visual Studio, as there are with CMake (search "multi_config_files") or with QMake.```

File: conans/test/functional/generators/visual_studio_multi_test.py (hunk 2 above)
Comment (SSE4) [reply conans/test/functional/generators/visual_studio_multi_test.py]: https://github.com/conan-io/conan/pull/4300/files#diff-bbfa689a2acdf6ea90ac7ab7b305af2aR128:
`client.run("export hello lasote/testing")`

**Comment by memsharded** [comment]
Hi @SSE4 

I have simplified the test and pushed it in https://github.com/conan-io/conan/pull/4300/commits/9b0149fa0c93d26d46f91e515ce7f82b8ac4412e

- The test was too slow, like 100 seconds, which is too much to add to the test suite for this scenario. 
- Building separate binaries for debug/release x86/64, does not add to the test suite, as the goal is precisely to be able to consume from one single package that is multi-config. I have simplified the re-used package to a header-only
- I am first creating the dependency, with "conan create", then later, testing the consumption. But in separate places (the second ``client.save()`` can remove the old files)
- I am checking the output of the .exe running, the see that the define is correctly processed.

**Pull request merged**